### ISSUE TITLE
Bootstrap Assets should us an established CDN service

### ISSUE THREAD

**Issue by mallardduck** [tuxnull/BootstraPHP#10]
**Describe the bug**
It appears that the main `bootstrap.min.css` file is being provided by using a URL that points to your domain. This will provide a poor user experience to developers and website end-users.

The reason being that this isn't an established CDN, so this raises two huge red flags. The first being that the chance of this providing a cache hit across multiple Bootstrap using sites is slim to none. Meaning the main benefit of using a CDN - higher chance of asset cache hits - is moot. And the other main benefit - files being delivered via regionally located servers - doesn't apply.

TL;DR: A domain behind CloudFlare !== a real asset CDN.

**Suggested Fix**
Utilize the official BootstrapCDN or CloudFlares CDNjs.

**Comment by tuxnull** [comment]
Hi. We have our own CDN which we use not only to always provide potential CSS fixes but also to ensure that Bootstrap doesn't just change their Stylesheets and render all BootstraPHP sites useless.

**Comment by mallardduck** [comment]
Guess I'm a little confused why you can't do that with the Bootstrap CDN.
It provides a versioned URL system so there shouldn't be a concern of them "chang[ing] their Stylesheets" and breaking things.

Unless you're shipping a customized version of bootstrap the same can be accomplished by using:

https://stackpath.bootstrapcdn.com/bootstrap/4.2.1/css/bootstrap.min.css

Even if you are shipping a customized bootstrap file I would argue for splitting customizations into a secondary CSS file. First load the stock `bootstrap.min.css` then load the file with modifications (and only modifications) directly after.

**Comment by tuxnull** [comment]
In our testing our CDN was faster than the Bootstrap CDN - maybe because we have less traffic. If this changes, I will reconsider switching to the Bootstrap CDN.

**Comment by mallardduck** [comment]
It feels as though you may be overlooking that every user (and internet connection) will have a different experience.

Here are the results I'm seeing:

### Without Cache:
|CDN Source | Run # | Time |
|---------- | -------- | -------|
| **Tuxnull** | 1 | 42ms |
| | 2 | 34ms |
| | 3 | 37ms |
| **BootstrapCDN** | 1 | 27ms |
| | 2 | 26ms |
| | 3 | 25ms |
| **CDNjs** | 1 | 24ms |
| | 2 | 20ms |
| | 3 | 33ms |

### With Cache:
|CDN Source | Run # | Time |
|---------- | -------- | -------|
| **Tuxnull** | 1 | 36ms |
| | 2 | 34ms |
| | 3 | 33ms |
| **BootstrapCDN** | 1 | 25ms |
| | 2 | 27ms |
| | 3 | 26ms |
| **CDNjs** | 1 | 18ms |
| | 2 | 17ms |
| | 3 | 16ms |

**Comment by tuxnull** [comment]
We tested it on multiple connections... Weird that your Tuxnull response time is worse than Cloudflare's since we use Cloudflare. Did you test it on actual content or just with ping?

**Comment by mallardduck** [comment]
Actual content - these tests were done in a real web browser. Tested with and without browser caching enabled via Chrome Dev tools. I'm not terribly surprised since CloudFlares CDNjs is easily more powerful than the free version (or lower tiers) of CloudFlare.

Further, this is only one half of the point I've raised. This won't be the best explanation, but here goes. The other half is the cache effcience you'd see from using a shared CDN. Using a shared CDN like CDNjs means there's a higher chance any given user will already have the file cached from visiting a different site.

So if I've already visiting a site utilzing Bootstrap 4 via CDNjs, then any site serving bootstrap from the same CDN gets to reuse this cached file. In the case of sites built using this tool my device will have to download and cache a second version of what's essnetially the same file.

**Comment by mallardduck** [comment]
As I said, not the best explanation. However I found an article that does a slightly better job explaining it using jQuery as an example instead of Bootstrap.

> The jQuery library will already be cached in the visitor's browser if they visited another website that references the same jQuery library. In this case, the visitor won't even have to download the jQuery library.

http://robertgreiner.com/2011/05/using-a-content-delivery-network-to-host-jquery/

**Comment by mallardduck** [comment]
@tuxnull Sorry, it seems I may not have come across as I expected. Wasn't trying to argue, or debate, about it to "win". Just wanted to provide my 2 cents based on my own experience.

**Comment by tuxnull** [comment]
Lol, i was just joking - I will implement it in ~ 30 minutes but it wont be in a full release until tomorrow.

**Comment by tuxnull** [comment]
Alright. init_stylesheet() now links the Bootstrap CDN and init_cdn_stylesheet() links the BootstraPHP CDN
